Use os.access in grains.core to check a file for usage

keesbos opened the pull request:
The os.path.exists(path) just checks for the existence of a file. The
os.access(path, os.R_OK) for the existence _and_ readability. The latter
is preferred in cases where that functionality is wanted.

Fixes a bug in 74508eb12555e462b5be8f7b36eea6fa56beb40a

See https://github.com/saltstack/salt/commit/74508eb12555e462b5be8f7b36eea6fa56beb40a#diff-6e894b39b67c4bab67796752744ea9d5R2044

### What does this PR do?
Fixes a bug for linux containers, where files exist, but are not readable for root.

### What issues does this PR fix or reference?

See https://github.com/saltstack/salt/commit/74508eb12555e462b5be8f7b36eea6fa56beb40a#diff-6e894b39b67c4bab67796752744ea9d5R2044

### Tests written?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

keesbos committed:
Use os.access in grains.core to check a file for usage

The os.path.exists(path) just checks for the existence of a file. The
os.access(path, os.R_OK) for the existence _and_ readability. The latter
is preferred in cases where that functionality is wanted.

Fixes a bug in 74508eb12555e462b5be8f7b36eea6fa56beb40a

See https://github.com/saltstack/salt/commit/74508eb12555e462b5be8f7b36eea6fa56beb40a#diff-6e894b39b67c4bab67796752744ea9d5R2044

salt-jenkins commented:
@keesbos, thanks for your PR! By analyzing the history of the files in this pull request, we identified @jfindlay, @gtmanfred and @secumod to be potential reviewers.

cachedout reviewed (changes requested):
@keesbos Thanks for this. However, it didn't pass the automated test suit. Could you please take a look?

keesbos commented:
I did see the failures for suse based systems. However, it doesn't make sense for me. They all throw this:
```
Traceback (most recent call last):
  File "/testing/tests/unit/grains/test_core.py", line 389, in test_suse_os_grains_tumbleweed
    self._run_suse_os_grains_tests(_os_release_map)
  File "/testing/tests/unit/grains/test_core.py", line 242, in _run_suse_os_grains_tests
    os_grains = core.os_data()
  File "/testing/salt/grains/core.py", line 1339, in os_data
    init_bin = salt.utils.which(init_cmdline[0])
IndexError: list index out of range
```
This is at:
```
            if _file_readable('/proc/1/cmdline'):
                with salt.utils.fopen('/proc/1/cmdline') as fhr:
                    init_cmdline = fhr.read().replace('\x00', ' ').split()
                    init_bin = salt.utils.which(init_cmdline[0])
```
Since it reaches this line, the return value of ```if _file_readable('/proc/1/cmdline'):``` is ```True```, which is correct (on my sles11 version I do have ```/proc/1/cmdline```). So far for the change in this PR.
On my sles11:
```
sles:~ # cat /etc/SuSE-release 
SUSE Linux Enterprise Server 11 (x86_64)
VERSION = 11
PATCHLEVEL = 3
sles:~ # cat /proc/1/cmdline ; echo
init [5]
```
So, the ```init_cmdline = fhr.read().replace('\x00', ' ').split()``` should result in something like ```['init', '[5]']```

So, I don't see the relation with this PR. I expect this is a test implementation error. Anyway, I'll try to see what the contents of ```/proc/1/cmdline``` is in these suse tests. Will create a separate (temporary) PR for this: https://github.com/saltstack/salt/pull/41228

secumod commented:
What about my comment in the source?

keesbos commented on salt/grains/core.py:
```diff
@@ -2041,7 +2045,7 @@ def _hw_data(osdata):
         for key, fw_file in sysfs_firmware_info.items():
             contents_file = os.path.join('/sys/class/dmi/id', fw_file)
-            if os.path.exists(contents_file):
+            if _file_readable(contents_file):
```
Technically, it would be best to this first:
```if osdata['kernel'] == 'Linux' and os.path.exists('/sys/class/dmi/id'):```
and enrich the collected data with other methods (as tested below) if possible/needed. I think that most of the time the other methods don't help much to collect data in containers.

keesbos commented:
A different approach is now in the develop branch that addresses the same problem (it works, but is not optimal). See commit 65be90ef21baa9ace86380c80f8cb30a9b7dcd27

BTW. It's a common error to use os.path.exists(path) where os.access(path, os.R_OK) should be used.

keesbos commented:
Fixed in https://github.com/saltstack/salt/pull/41249
